modules/OFConnectionManager2/utest/main.c

```diff
@@ -512,6 +512,37 @@ tcp_recvmsg(int tl, uint8_t buf[], int buflen,
     return of_object_new_from_message_preallocated(storage, buf, msglen);
 }
 
+static of_object_t *
+tcp_no_recvmsg(int tl, uint8_t buf[], int buflen,
+               of_object_storage_t *storage)
+{
+    const int header_bytes = 8;
+    int ret = 0;
+    uint16_t msglen;
+    int i = 0;
+
+    do {
+        ret = read(tl, buf, header_bytes);
+        i++;
+    } while (i < 256 && ret == -1 && errno == EAGAIN);
+    if (ret == -1 && errno == EAGAIN) {
+        return NULL;
+    } else {
+        INDIGO_ASSERT(ret != -1, "read error: %s", strerror(errno));
+        INDIGO_ASSERT(ret == header_bytes, "read %d bytes, expected %d", 
+                  ret, header_bytes);
+        msglen = ntohs(*((uint16_t*)(buf+2)));
+        INDIGO_ASSERT(msglen <= buflen, "message len %d exceeds buffer len %d",
+                  msglen, buflen);
+        ret = read(tl, buf+header_bytes, msglen-header_bytes);
+        INDIGO_ASSERT(ret == msglen-header_bytes, 
+                  "body read %d bytes, expected %d",
+                  ret, msglen-header_bytes);
+
+        return of_object_new_from_message_preallocated(storage, buf, msglen);
+    }
+}
+
```

Comment:
to avoid creating lots of similar code that will be harder to maintain in the long term, let's refactor tcp_recvmsg and tcp_no_recvmsg into a single function whose behavior is controlled by input parameters.